Move all unit-test-code to unit subdirectory under tests.

petervdmeer opened the pull request:
**Feature:** This PR moves all code and support files for testing to their own (sub-)directories

## Feature Details
The code for unit testing was mixed with the scripts for integration-tests and with one-script-checkers. This would make it difficult for anyone starting with new tests to find out which script is for what.

This PR moves the tests all to their own sub-directories and adds some README.md files to allow new test-writers to onboard more easily.

## UI Screenshots
N/A

## Usage Examples
N/A

## Testing Done
Performed some testing locally (unit tests through scons, executed some checkers, executed integration tests non-headless). 

## Performance Impact
N/A

quyykk reviewed (approved):
LGTM! But I'd wait for tehhowch again.

tehhowch commented on SConstruct:
```diff
@@ -164,12 +164,13 @@ env.Default(sky)
 # The testing infrastructure ignores "mode" specification (i.e. we only test optimized output).
 # (If we add support for code coverage output, this will likely need to change.)
 testBuildDirectory = pathjoin("tests", env["BUILDDIR"])
```
note that you're still building the test object files and binary into `tests/build/*.o` and `tests/endless-sky-tests`, respectively

petervdmeer replied:
Should be fixed with the the last updates I did.

tehhowch reviewed (commented):
When adding the framework, I had picked the locations & names of `src` and `include` to be somewhat aligned with what one might expect for compiled tests.

I think if we're looking to reduce confusion & uncertainty with writing tests for Endless Sky, we are better off adding a README to the `tests/` directory, and moving the not-compiled-tests to more suitable places.

What about:
 - "project checkers" and "parse-test drivers" move from `tests/` to `utils/` (we expect all utility scripts to be consumable by users, and these scripts are)
 - "drive the game" scripts & support into a subfolder or sibling, bringing `config/` along, too
   - I know we want to be able to set up per-headless-test config / data, so we should keep that in mind when reorganizing the folders.

quyykk commented on utils/check_codeblocks.sh:
```diff
@@ -26,7 +26,7 @@ do
 done
 
-for FILE in $(find tests/src/ -type f -name "*.h" -o -name "*.cpp" | sed s,^tests/src/,, | sort)
+for FILE in $(find tests/unit/ -type f -name "*.h" -o -name "*.cpp" | grep -v -e "^tests/unit/include" | sed s,^tests/unit/src,, | sort)
```
You don't need the additional grep filter if you restrict searching the first find to `tests/unit/src/`.

petervdmeer commented:
> When adding the framework, I had picked the locations & names of `src` and `include` to be somewhat aligned with what one might expect for compiled tests.

Yes, that is nice (and nicer than my intermediate version where I just dumped them in the src directory). I re-created a src-directory under `tests/unit` where I placed all the cpp again. But I did move the README.md file one directory higher, because I feel that is where developers will look for it.

> I think if we're looking to reduce confusion & uncertainty with writing tests for Endless Sky, we are better off adding a README to the `tests/` directory, and moving the not-compiled-tests to more suitable places.

Good idea. I added a generic README.md and modified the ones for the unit-tests and for the integration-tests.

> What about:
> 
>     * "project checkers" and "parse-test drivers" move from `tests/` to `utils/` (we expect all utility scripts to be consumable by users, and these scripts are)

Done. (And I mentioned their locations in the README.md file, just in case somebody is looking for them.)

>     * "drive the game" scripts & support into a subfolder or sibling, bringing `config/` along, too

Done by merging my other PR into this one.

>       * I know we want to be able to set up per-headless-test config / data, so we should keep that in mind when reorganizing the folders.

Yes, this is the reason why I was actually looking into those directories. I added information in the integration/README.md file on the directory I plan to use for this.

petervdmeer commented on EndlessSkyTests.cbp:
```diff
@@ -50,7 +50,8 @@
 			<Add option="-fno-rtti" />
-			<Add directory="tests/include" />
+			<Add directory="tests/unit/include" />
+			<Add directory="tests/unit" />
```
I agree, this was a bit ugly. I updated the paths in all affected unit-test-files instead.

petervdmeer commented:
Merging this (approved by quyykk and all open review-comments from tehhowch are handled); there is a discussion on Discord from code-contributors that are told (by me) that they should write tests. I don't want to make them wait for this PR.

There will be follow-up work on the tests (like the test-universes), so if there still are opens on this PR that we find after the PR, then we will include them in the follow-up work.
